### PR TITLE
Alternate Thread Creation API To Avoid Memory Leaks

### PR DESCRIPTION
**Pull request by NoMoreFood**
- Switched from CreateThread() to _beginthreadex() and ExitThread() to _endthreadex() in order to avoid potential leaks when linking with static CRT library.
- Addressed a variety of warnings that were being detected with static code analysis.

### CODE REVIEW

File: contrib/win32/win32compat/Debug.h [hunk 1]
```diff
@@ -1,6 +1,6 @@
 #pragma once
 
-void     fatal(const char *, ...);
+void     __declspec(noreturn) fatal(const char *, ...);
```
Comment (bagajjal) [inline contrib/win32/win32compat/Debug.h:3]: why is the special treatment for fatal() but not for error/verbose/debug?

Comment (bagajjal) [reply contrib/win32/win32compat/Debug.h:3]: IntelliSense doesn't show these warnings.. which static analysis tool, you used?

File: contrib/win32/win32compat/misc.c [hunk 2]
```diff
@@ -931,7 +931,7 @@ resolved_path_utf16(const char *input_path)
 		if (changed_req > 0) {
 			wchar_t * resolved_path_new = realloc(resolved_path, 
 				(resolved_len + changed_req + 1) * sizeof(wchar_t));
-			if (resolved_path == NULL) {
+			if (resolved_path_new == NULL) {
```
Comment (bagajjal) [inline contrib/win32/win32compat/misc.c]: good catch :)

**Review by bagajjal — CHANGES_REQUESTED** [review]

**Review by bagajjal — APPROVED** [review]

File: contrib/win32/win32compat/Debug.h (hunk 1 above)
Comment (NoMoreFood) [reply contrib/win32/win32compat/Debug.h:3]: The fatal() function is the only call that will force the program to exit.  The others will not. The declaration informs the compiler the function will never return and will use this information to optimize compilation as well as assist in static code analysis (e.g., detecting unreachable code, etc).

In Visual Studio, you can do Analyze -> Run Code Analysis -> Run Code Analysis on [project name].  However, the results aren't always perfect, particularly when it comes to checking bounds on dynamic memory allocations that are based on runtime-calculated values.

File: contrib/win32/win32compat/misc.c (hunk 2 above)
Comment (NoMoreFood) [reply contrib/win32/win32compat/misc.c]: :-) Code analysis actually caught this.  It noted the fact that resolved_path_new was never checked for a null value prior to use.

**Comment by bingbing8** [comment]
close and reopen to trigger appveyour build report

File: contrib/win32/win32compat/misc.c [hunk 3]
```diff
@@ -1052,10 +1052,10 @@ resolved_path_utf16(const char *input_path)
...
 			wchar_t * resolved_path_new = realloc(resolved_path, 
 				(resolved_len + changed_req + 1) * sizeof(wchar_t));
-			if (resolved_path == NULL) {
-				debug3("%s: memory allocation failed.", __FUNCTION__);
-				free(resolved_path);
-				errno = ENOMEM;
+			if (resolved_path_new == NULL) {
+				debug3("%s: memory allocation failed.", __FUNCTION__);
+				free(resolved_path);
+        errno = ENOMEM;
```
Comment (manojampalam) [inline contrib/win32/win32compat/misc.c]: fix indentation

**Review by manojampalam — CHANGES_REQUESTED** [review]
Please merge in the latest changes from latestw_all and resolve conflicts. 
Also open an issue to push up warning level in posix_compat project to "Level 4". These should have been automatically caught during regular builds.

**Pull request merged**